File: packages/format/test/vp9-superframe.test.ts
import { describe, expect, it } from "vitest";

import { splitVp9Superframe } from "../src/vp9/superframe.js";

describe("VP9 superframe parsing", () => {
  it("splits hidden and displayed coded frames into owned bytes", () => {
    const marker = 0xc1;
    const packet = Uint8Array.of(0x84, 0xaa, 0x86, 0xbb, 0xcc, marker, 2, 3, marker);
    const frames = splitVp9Superframe(packet);
    expect(frames).toEqual([
      Uint8Array.of(0x84, 0xaa),
      Uint8Array.of(0x86, 0xbb, 0xcc)
    ]);
    packet.fill(0);
    expect(frames[0]).toEqual(Uint8Array.of(0x84, 0xaa));
  });

  it("returns a detached single frame and rejects malformed indexes", () => {
    const packet = Uint8Array.of(0x86, 0x01);
    const frames = splitVp9Superframe(packet);
    packet.fill(0);
    expect(frames).toEqual([Uint8Array.of(0x86, 0x01)]);
    expect(() => splitVp9Superframe(Uint8Array.of(0x84, 0xc1, 1, 1, 0xc1)))
      .toThrow(/sizes/u);
  });
});

File: flutter/packages/aval_format/test/vp9_superframe_test.dart
/// Dart port of `packages/format/test/vp9-superframe.test.ts`.
library;

import 'dart:typed_data';

import 'package:aval_format/src/errors.dart';
import 'package:aval_format/src/vp9/index.dart';
import 'package:test/test.dart';

void main() {
  group('VP9 superframe parsing', () {
    test('splits hidden and displayed coded frames into owned bytes', () {
      const marker = 0xc1;
      final packet = Uint8List.fromList(
          [0x84, 0xaa, 0x86, 0xbb, 0xcc, marker, 2, 3, marker]);
      final frames = splitVp9Superframe(packet);
      expect(frames, equals([
        Uint8List.fromList([0x84, 0xaa]),
        Uint8List.fromList([0x86, 0xbb, 0xcc]),
      ]));
      packet.fillRange(0, packet.length, 0);
      expect(frames[0], equals(Uint8List.fromList([0x84, 0xaa])));
    });

    test('returns a detached single frame and rejects malformed indexes', () {
      final packet = Uint8List.fromList([0x86, 0x01]);
      final frames = splitVp9Superframe(packet);
      packet.fillRange(0, packet.length, 0);
      expect(frames, equals([Uint8List.fromList([0x86, 0x01])]));
      expect(
        () => splitVp9Superframe(Uint8List.fromList([0x84, 0xc1, 1, 1, 0xc1])),
        throwsA(predicate<Object>((error) =>
            error is FormatError && RegExp('sizes').hasMatch(error.message))),
      );
    });
  });
}
